py/client/pydeephaven/session.py

```diff
@@ -338,38 +391,83 @@ def _connect(self):
+            # from the server which will contain the bearer token we will
+            # use in subsequent RPCs; the token will be included in the updated
+            # value for self._auth_header_value that will happen through a call
+            # to update_metadata.
             config_dict = self.config_service.get_configuration_constants()
             session_duration = config_dict.get("http.session.durationMs")
             if not session_duration:
                 raise DHError("server configuration is missing http.session.durationMs")
 
+            self._timeout_seconds = int(session_duration.string_value)/1000.0
+            # Random skew to ensure multiple processes that may have
+            # started together don't align retries.
+            skew = random()
+            # Backoff schedule for retries after consecutive failures to refresh auth token
+            self._refresh_backoff = [ skew + 0.1, skew + 1, skew + 10 ]
+            
+            if self._refresh_backoff[0] > self._timeout_seconds:
+                raise DHError(f'server configuration http.session.durationMs={session_duration} is too small.')
+            if 0.25*self._timeout_seconds < self._refresh_backoff[-1]:
+                self._refresh_backoff.extend(
+                    [skew + 0.25 * self._timeout_seconds,
+                     skew + 0.35 * self._timeout_seconds,
+                     skew + 0.45 * self._timeout_seconds])
+            for i in range(1, len(self._refresh_backoff)):
+                if self._refresh_backoff[i] > self._timeout_seconds:
+                    self._refresh_backoff = self._refresh_backoff[0:i]
+                    break
+
             self.is_connected = True
 
-            self._timeout = int(session_duration.string_value)
             if self._never_timeout:
                 self._keep_alive()
 
     def _keep_alive(self):
-        if self.is_connected:
-            if self._keep_alive_timer:
-                self._refresh_token()
-            self._keep_alive_timer = threading.Timer(self._timeout / 2 / 1000, self._keep_alive)
-            self._keep_alive_timer.daemon = True
-            self._keep_alive_timer.start()
-
-    def _refresh_token(self):
-        try:
-            self._flight_client.authenticate(self._auth_handler)
-        except Exception as e:
+        trace(f'_keep_alive')
+        if not self.is_connected:
+            return
+        ok = True
+        if self._keep_alive_timer:
+            ok = self._refresh_token()
+            if ok:
+                self._refresh_failures = 0
+            else:
+                self._refresh_failures += 1
+        if self._refresh_failures == 0:
+            timer_wakeup = 0.5*self._timeout_seconds
+        elif self._refresh_failures >= len(self._refresh_backoff):
             self.is_connected = False
-            raise DHError("failed to refresh auth token") from e
+            raise DHError(f'Failed to refresh token {self._refresh_failures} times')
```

Comment:
To confirm, what happens if this throws on the keepalive thread - does the user get some notification? I think they'll still have the old auth token, so might not notice until some other call goes off..?

Might make more sense to do a warning log here and just return, since nothing can ever catch this? Sort of as an alternative to the logging we will now miss on line 459.

Reply:
> what happens if this throws on the keepalive thread 
I thought the program would end, but that's not the case.  So I am going to call os.exit(1) and log critical.